azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java

```diff
@@ -617,4 +622,41 @@ public static void deleteSymlinkedFile(final Path symlinkedFilePath)
       throw new ExecutorManagerException(e);
     }
   }
+
+  /**
+   * Shutdown the Container. This shuts down the ExecutorService which runs the flow execution
+   * as well as JettyServer.
+   */
+  private void shutdown() {
+    logger.info("Shutting down the pod");
+    while (!this.flowFuture.isDone()) {
+      // This should not happen immediately as submitFlowRunner is a blocking call.
+      try {
+        Thread.sleep(100);
+      } catch (final InterruptedException e) {
+        logger.error("The sleep while waiting for execution : {} to finish was interrupted",
+                this.flowRunner.getExecutionId());
+      }
+    }
+
+    try {
+      this.executorService.shutdown();
+      boolean result = false;
+      while (!result) {
+        logger.info("Awaiting Shutdown of executing flow");
+        try {
+          result = this.executorService.awaitTermination(5, TimeUnit.SECONDS);
```

Comment:
Will this block indefinitely due to the `while` loop if the the executorService is stuck for some reason ?
Consider removing the while loop and log an error if the executorService is still not shutdown, making sure the rest of the procedure can still complete and cleanup any resources (system.exit should do that, worth confirming though)

Reply:
Thanks for pointing this out. I intended to limit the while loop.
Eventually, all this is not necessary as System.exit will just wipe everything out and pod completion will cleanup the pod too.
Nonetheless, as a good practice, I will limit the loop.